values/src/main/java/org/spine3/net/UrlParser.java

```diff
@@ -0,0 +1,173 @@
+        }
+
+        record.setProtocol(protocolBuilder.build());
+    }
+
+    private void parseCredentials() {
+        final int credentialsEndingIndex = unProcessedInput.indexOf(CREDENTIALS_ENDING);
+        if (credentialsEndingIndex == -1) {
+            return;
+        }
+
+        final String credential = unProcessedInput.substring(0, credentialsEndingIndex);
+        unProcessedInput = unProcessedInput.substring(credentialsEndingIndex + CREDENTIALS_ENDING.length());
+
+        final Url.Record.Authorization.Builder auth = Url.Record.Authorization.newBuilder();
+
+        final int credentialsSeparatorIndex = credential.indexOf(CREDENTIALS_SEPARATOR);
+        if (credentialsSeparatorIndex != -1) {
+            final String user = credential.substring(0, credentialsSeparatorIndex);
+            final String password = credential.substring(
+                    credentialsSeparatorIndex + CREDENTIALS_SEPARATOR.length());
+            auth.setPassword(password);
+            auth.setUser(user);
+        } else {
+            auth.setUser(credential);
+        }
+
+        record.setAuth(auth.build());
+    }
+
+    private void parseHost() {
+        final int hostEndingIndex = unProcessedInput.indexOf(HOST_ENDING);
+        final String host;
+
+        if (hostEndingIndex == -1) {
+            host = unProcessedInput;
+            unProcessedInput = "";
+        } else {
+            host = unProcessedInput.substring(0, hostEndingIndex);
+            unProcessedInput = unProcessedInput.substring(hostEndingIndex + HOST_ENDING.length());
+        }
+
+        final int portIndex = host.indexOf(HOST_PORT_SEPARATOR);
+        if (portIndex != -1) {
+            final String port = host.substring(portIndex + HOST_PORT_SEPARATOR.length());
+            record.setPort(port);
+            final String hostAddress = host.substring(0, portIndex);
+            record.setHost(hostAddress);
+        } else {
+            record.setHost(host);
+        }
+    }
+
+    private void parseFragment() {
+        final int fragmentIndex = unProcessedInput.lastIndexOf(FRAGMENT_START);
+        if (fragmentIndex == -1) {
+            return;
+        }
+
+        final String fragment = unProcessedInput.substring(fragmentIndex + FRAGMENT_START.length());
+        unProcessedInput = unProcessedInput.substring(0, fragmentIndex);
+
+        record.setFragment(fragment);
+    }
+
+    private void parseQueries() {
+        final int queriesStartIndex = unProcessedInput.indexOf(QUERIES_START);
+        if (queriesStartIndex == -1) {
+            return;
+        }
+
+        final String queriesString = unProcessedInput.substring(queriesStartIndex + QUERIES_START.length());
+        unProcessedInput = unProcessedInput.substring(0, queriesStartIndex);
+
+        final String[] queries = queriesString.split(QUERIES_SEPARATOR);
+        for (String query : queries) {
+            // TODO:2016-05-23:mikhail.mikhaylov: Split arguments if needed.
```

Comment:
Please address the TODO. See my earlier comment as to the structured type for queries.